ci/kind/kind-setup.sh

```diff
@@ -528,19 +528,28 @@ function destroy_external_servers {
 
 function clean_kind {
     echo "=== Cleaning up stale kind clusters ==="
-    read -a all_kind_clusters <<< $(kind get clusters)
-    for kind_cluster_name in "${all_kind_clusters[@]}"; do
-        creationTimestamp=$(kubectl get nodes --context kind-$kind_cluster_name -o json -l node-role.kubernetes.io/control-plane | \
-        jq -r '.items[0].metadata.creationTimestamp')
+    LOCK_FILE="$THIS_DIR/.kube/config.lock"
+    (
+      flock -x 200
+
+      for context in $(kubectl config get-contexts -o name | grep 'kind-'); do
+        cluster_name=$(echo $context | sed 's/^kind-//')
+        creationTimestamp=$(kubectl get nodes --context $context -o json -l node-role.kubernetes.io/control-plane | \
+        jq -r '.items[0].metadata.creationTimestamp' || true)
+        if [[ -z $creationTimestamp ]]; then
+          continue
+        fi
         creation=$(printUnixTimestamp "$creationTimestamp")
         now=$(date -u '+%s')
         diff=$((now-creation))
         timeout=$(($UNTIL_TIME_IN_MINS*60))
         if [[ $diff -gt $timeout ]]; then
-           echo "=== kind ${kind_cluster_name} present from more than $UNTIL_TIME_IN_MINS minutes ==="
-           kind delete cluster --name $kind_cluster_name
+          echo "=== Cluster ${cluster_name} is present for more than $UNTIL_TIME_IN_MINS minutes ==="
+          kind delete cluster --name $cluster_name
         fi
-    done
+      done
+    )200>>"$LOCK_FILE"
+    rm -rf $LOCK_FILE
```

Comment:
this is not a good idea IMO. It feels like there can be a race condition where we delete the file while another job is holding the lock?

Reply:
But untill we delete config.lock the other process will not be able to acquire the lock and it may panic.
Also since flock is not compatible with the Go based locking mechanism so what's you opinion on that should we use the alternative approach of writing the cluster names to a file and acquiring lock over that, or. should we move ahead with acquiring lock over the kubeconfig and use the below approach:

before invoking `kind create cluster` command we can use flock to wait for the other process to release the lock and then we should trigger this command so in that case flock will not interfere with the Go based locking mechanism its just that we will be introducing couple of unnecessary flocks in the code.

Reply:
We should not use `~/.kube/config.lock`, I think that's a given. We can pretend it doesn't exist.
So we need our own lock file. After that, we have 2 options, and you can choose which one you want to use:
1) have our own state file to keep track of cluster names and creation timestamps (which I was originally proposing)
2) only rely on kubectl / kind, and do not introduce our own state file. With this option we use flock to acquire a lock before calling kubectl / kind, as appropriate

I think both approaches will work

Reply:
No we don't rely on `~/.kube/config.lock` for either approach, as we have discussed about why this is not a viable option.

Reply:
okay

Reply:
Updated the code as per the requirement.
An important note:
After this PR is merged we will have to update all the kind testbeds to include the .clusters file in the home directory which is "/var/lib/jenkins" or else the other running kind test may fail.

@antoninbas can you review the PR once.
Thanks for the suggestions.

Reply:
If the `.clusters` file is missing, ideally this should be handled gracefully, i.e., we should treat it the same way as an empty file

Reply:
I think we do not need to handle that explicitly echo will take care of that.

Reply:
I'm referring to your "important note" above:

> After this PR is merged we will have to update all the kind testbeds to include the .clusters file in the home directory which is "/var/lib/jenkins" or else the other running kind test may fail.

I don't think we should have that requirement for manual update. The script should handle this gracefully. Is that not the case?

Reply:
Yes it will handle it, the. note is just for the testbeds where /var/lib/jenkins directory is missing

Reply:
ideally we have it on all our testbeds so this should not be an issue @XinShuYang and @KMAnju-2021 can confirm.

Reply:
Yes, iirc we have this /var/lib/jenkins directory on all kind testbeds.

Reply:
Actually it's good that you highlight this. `/var/lib/jenkins/` is not an appropriate directory to use here as `ci/kind/kind-setup.sh` is used for local development, Github CI, etc. Please use a more "universal" directory, such as `~/.kube/antrea/` (you can create the directory if it doesn't exist). I guess `~/.antrea/` would also be a good choice if we don't want to write anything to the `~/.kube` directory.

Reply:
I guess ~/.antrea would be a better option here